### PR TITLE
POC: In Memory Repository Refactor

### PR DESCRIPTION
**Pull request by robertbennett1998**
Initial rough design:
image

### CODE REVIEW

File: Microsoft.Azure.CosmosRepository.InMemory/IDbInterface.cs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿namespace Microsoft.Azure.CosmosRepository.InMemory;
```
Comment (robertbennett1998) [inline Microsoft.Azure.CosmosRepository.InMemory/IDbInterface.cs:1]: Needs deleting

File: Microsoft.Azure.InMemoryCosmosRepositoryAcceptanceTests/CosmosRepositoryInMemoryAcceptanceTest.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
+// Copyright (c) IEvangelist. All rights reserved.
```
Comment (robertbennett1998) [inline Microsoft.Azure.InMemoryCosmosRepositoryAcceptanceTests/CosmosRepositoryInMemoryAcceptanceTest.cs:1]: Needs deleting

File: Microsoft.Azure.InMemoryCosmosRepositoryAcceptanceTests/Microsoft.Azure.InMemoryCosmosRepositoryAcceptanceTests.csproj [hunk 3]
```diff
@@ -0,0 +1,29 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (robertbennett1998) [inline Microsoft.Azure.InMemoryCosmosRepositoryAcceptanceTests/Microsoft.Azure.InMemoryCosmosRepositoryAcceptanceTests.csproj:1]: Needs deleting. Apparently rider didnt want to delete the actual files.

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Writer/IItemStoreWriterStrategy.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            {
+                Type genericType = typeof(IItemStoreWriterStrategyStep<>).MakeGenericType(type);
+                services.AddRange(_serviceProvider.GetServices(genericType)
+                    .Select(service => (IItemStoreWriterStrategyStep)service));
+            }
+
+            return services;
+        }
+
+        public async ValueTask<string> WriteAsync(TItem item, string id, string? partitionKey = null)
```
Comment (robertbennett1998) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Writer/IItemStoreWriterStrategy.cs:45]: Rename to transform async

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/IItemStore.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace Microsoft.Azure.CosmosRepository.InMemory
+{
+    internal interface IItemStore<TItem> where TItem : IItem
+    {
+        ValueTask<TItem> WriteAsync(TItem item, string id, string? partitionKey = null);
+        ValueTask<TItem> ReadAsync(string id, string? partitionKey = null);
+    }
+
+    internal class ItemStore<TItem> : IItemStore<TItem> where TItem : IItem
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/IItemStore.cs:17]:
```suggestion
    internal sealed class ItemStore<TItem> : IItemStore<TItem> where TItem : IItem
```

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/IItemStore.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            partitionKey ??= id;
+            string json = await _writerStrategy.WriteAsync(item, id, partitionKey);
+
+            if (_itemStore.TryGetValue(partitionKey, out ConcurrentDictionary<string, string>? partitionStore))
+            {
+                partitionStore[id] = json;
+            }
+            else
+            {
+                _itemStore[partitionKey] = new ConcurrentDictionary<string, string>
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/IItemStore.cs:39]:
```suggestion
                _itemStore[partitionKey] ??= new ConcurrentDictionary<string, string>
```

It's possible that with concurrent writes that two threads could evaluate line 33 at the same time and both end up in the else. At which time they'd both create a new instance, missing the first one. Seems like a race condition.

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/IItemStore.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                };
+            }
+
+            return await ReadAsync(id, partitionKey);
+        }
+
+        public ValueTask<TItem> ReadAsync(string id, string? partitionKey = null)
+        {
+            throw new NotImplementedException();
+        }
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/IItemStore.cs:51]:
```suggestion
        public ValueTask<TItem?> ReadAsync(string id, string? partitionKey = null)
        {
            if (partitionKey is not null && 
                _itemStore.ContainsKey(partitionKey))
            {
                return _itemStore[partitionKey].TryGetValue(id, out var item)
                    ? item
                    : null;
            }
            
            return null;
        }
```

Wouldn't this need to be nullable?

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Writer/IItemStoreWriterStrategy.cs (hunk 4 above)
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Writer/IItemStoreWriterStrategy.cs:45]:
```suggestion
        public async ValueTask<string> TransformAsync(TItem item, string id, string? partitionKey = null)
```

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Writer/IItemStoreWriterStrategy.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.CosmosRepository.InMemory
+{
+    internal interface IItemStoreWriterStrategy<TItem> where TItem : IItem
+    {
+        ValueTask<string> WriteAsync(TItem item, string id, string? partitionKey = null);
+    }
+
+    internal class ItemStoreWriterStrategy<TItem> : IItemStoreWriterStrategy<TItem> where TItem : IItem
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Writer/IItemStoreWriterStrategy.cs:21]:
```suggestion
    internal sealed class ItemStoreWriterStrategy<TItem> : IItemStoreWriterStrategy<TItem> where TItem : IItem
```

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Writer/IItemStoreWriterStrategy.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+using Microsoft.Azure.CosmosRepository.Extensions;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.CosmosRepository.InMemory
+{
+    internal interface IItemStoreWriterStrategy<TItem> where TItem : IItem
+    {
+        ValueTask<string> WriteAsync(TItem item, string id, string? partitionKey = null);
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Writer/IItemStoreWriterStrategy.cs:18]:
```suggestion
        ValueTask<string> TransformAsync(TItem item, string id, string? partitionKey = null);
```

**Review by IEvangelist — APPROVED** [review]
This awesome, great job. Let's clean it up a bit and :shipit:

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/ItemStore.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        private readonly IItemStoreWriterStrategy<TItem> _writerStrategy;
+        private readonly IItemStoreReaderStrategy<TItem> _readerStrategy;
+        private readonly IOptionsMonitor<RepositoryOptions> _optionsMonitor;
+
+        public ItemStore(
+            IItemStoreWriterStrategy<TItem> writerStrategy,
+            IItemStoreReaderStrategy<TItem> readerStrategy,
+            IOptionsMonitor<RepositoryOptions> optionsMonitor
+            )
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/ItemStore.cs:24]:
```suggestion
            IOptionsMonitor<RepositoryOptions> optionsMonitor)
```

File: src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Reader/IItemStoreReaderStrategy.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.CosmosRepository.InMemory.Reader
+{
+    internal interface IItemStoreReaderStrategy<TItem> where TItem : IItem
+    {
+        ValueTask<TItem> TransformAsync(JObject itemJObject);
+
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/Repositories/InMemory/Reader/IItemStoreReaderStrategy.cs:12]:
```suggestion
```

**Review by IEvangelist — COMMENTED** [review]
This looks good!

**Comment by IEvangelist** [comment]
Closing all open PRs that have conflicts, build validation errors, and haven't been updated in more than a month.

**Pull request closed without merging**